REGRESSION(264795@main): Some HLS videos have a frame half as wide as it should be

aestes opened the pull request:
#### a517dfc6d1cbd03b09f5736cf4b1bf62f8e5ea88
<pre>
REGRESSION(264795@main): Some HLS videos have a frame half as wide as it should be
<a href="https://bugs.webkit.org/show_bug.cgi?id=260825">https://bugs.webkit.org/show_bug.cgi?id=260825</a>
rdar://110467872

Reviewed by Jer Noble.

In 264795@main, the source of truth for videoLayerSize was moved from MediaPlayerPrivate to
HTMLMediaElement, but a cached value remained on MediaPlayerPrivateRemote (in m_videoLayerSize).
When the GPU process would send the LayerHostingContextIdChanged message to the WebContent process
with a new presentation size, MediaPlayerPrivateRemote&apos;s cached value would be updated but
HTMLMediaElement&apos;s would not. This could ultimately lead to WebAVPlayerLayer laying out with a size
that didn&apos;t match the presentation size of the styled video element.

Fixed this by removing m_videoLayerSize from MediaPlayerPrivateRemote and delegating to
HTMLMediaElement when setting the presentation size specified in the LayerHostingContextIdChanged
message.

No new tests as reproducing this requires content that isn&apos;t licenced for inclusion in WebKit.

* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::videoLayerSizeDidChange):
* Source/WebCore/platform/graphics/MediaPlayer.h:
(WebCore::MediaPlayerClient::mediaPlayerVideoLayerSizeDidChange):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::createAVPlayerLayer):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::ensureLayer):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::platformLayer const):
(WebKit::MediaPlayerPrivateRemote::setVideoLayerSizeFenced): Deleted.
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/cocoa/MediaPlayerPrivateRemoteCocoa.mm:
(WebKit::MediaPlayerPrivateRemote::layerHostingContextIdChanged):
(WebKit::MediaPlayerPrivateRemote::videoLayerSize const):
(WebKit::MediaPlayerPrivateRemote::setVideoLayerSizeFenced):

Canonical link: <a href="https://commits.webkit.org/267389@main">https://commits.webkit.org/267389@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1aeefac3dcae17d0745be5e54c33c900c530926f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16450 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16771 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17207 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18226 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15425 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16640 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20002 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16914 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17782 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16646 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17069 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14226 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18997 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14313 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14910 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21702 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15301 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15075 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19382 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15664 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13302 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14866 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3944 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19235 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15490 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebKit/WebProcess/GPU/media/cocoa/MediaPlayerPrivateRemoteCocoa.mm:
```diff
@@ -92,7 +92,20 @@
+    RefPtr player = m_player.get();
+    if (!player)
+        return { };
+    return player->videoLayerSize();
```
Seems like this could be:
```
if (RefPtr player = m_player.get())
    return player->videoLayerSize();
return { };
```

jernoble commented on Source/WebCore/platform/graphics/MediaPlayer.cpp:
```diff
@@ -923,6 +923,11 @@ FloatSize MediaPlayer::videoLayerSize() const
+{
+    client().mediaPlayerDidSetVideoLayerSize(size);
+}
+
```
`MediaPlayer` is a weird class that both passes information down to MediaPlayerPrivate, up to MediaPlayerClient, and also does some amount of caching. That said, the naming of this method seems wrong.

Maybe this could be `MediaPlayer::videoLayerSizeDidChange(const FloatSize& size)`?

aestes committed:
REGRESSION(264795@main): Some HLS videos have a frame half as wide as it should be
https://bugs.webkit.org/show_bug.cgi?id=260825
rdar://110467872

Reviewed by Jer Noble.

In 264795@main, the source of truth for videoLayerSize was moved from MediaPlayerPrivate to
HTMLMediaElement, but a cached value remained on MediaPlayerPrivateRemote (in m_videoLayerSize).
When the GPU process would send the LayerHostingContextIdChanged message to the WebContent process
with a new presentation size, MediaPlayerPrivateRemote's cached value would be updated but
HTMLMediaElement's would not. This could ultimately lead to WebAVPlayerLayer laying out with a size
that didn't match the presentation size of the styled video element.

Fixed this by removing m_videoLayerSize from MediaPlayerPrivateRemote and delegating to
HTMLMediaElement when setting the presentation size specified in the LayerHostingContextIdChanged
message.

No new tests as reproducing this requires content that isn't licenced for inclusion in WebKit.

* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::MediaPlayer::videoLayerSizeDidChange):
* Source/WebCore/platform/graphics/MediaPlayer.h:
(WebCore::MediaPlayerClient::mediaPlayerVideoLayerSizeDidChange):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::createAVPlayerLayer):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::ensureLayer):
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::platformLayer const):
(WebKit::MediaPlayerPrivateRemote::setVideoLayerSizeFenced): Deleted.
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/cocoa/MediaPlayerPrivateRemoteCocoa.mm:
(WebKit::MediaPlayerPrivateRemote::layerHostingContextIdChanged):
(WebKit::MediaPlayerPrivateRemote::videoLayerSize const):
(WebKit::MediaPlayerPrivateRemote::setVideoLayerSizeFenced):

Canonical link: https://commits.webkit.org/267389@main
